libs/nest/config/src/lib/ConfigurationLoader.ts

```diff
@@ -0,0 +1,162 @@
+    if (error) {
+      throw error
+    }
+
+    if (result === undefined) {
+      throw new ConfigurationError(
+        `Failed loading configuration for ${this.definition.name}. Config was undefined.`,
+      )
+    }
+
+    if (this.isDev && hasMissingIssue) {
+      logger.warn({
+        message: `Could not load configuration for ${this.definition.name}. Missing ${issues.length} required environment variable(s).`,
+        category: 'ConfigModule',
+      })
+      return new InvalidConfiguration(errorMessage) as Configuration<T>
+    }
+
+    this.validateSchema(result)
+
+    return {
+      ...result,
+      isConfigured: true,
+    }
+  }
+
+  private validateSchema(result: T) {
+    if (!this.definition.schema) {
+      return
+    }
+
+    const parseResult = this.definition.schema.safeParse(result)
+
+    if (!parseResult.success) {
+      throw new ConfigurationError(
+        this.formatValidationErrors(parseResult.error),
+      )
+    }
+  }
+
+  private formatIssues(issues: Array<[string, Issue]>) {
+    const formattedIssues = issues
+      .map((issue) => `- ${issue[0]} ${issue[1]}`)
+      .join('\n')
+    return `Failed loading configuration for ${this.definition.name}:\n${formattedIssues}`
+  }
+
+  private formatValidationErrors(error: ZodError) {
+    const formattedErrors = error.errors
+      .map((error) => `- ${error.path.join('.') || '<ROOT>'}: ${error.message}`)
+      .join('\n')
+    return `Failed loading configuration for ${this.definition.name}. Validation failed:\n${formattedErrors}`
+  }
+
+  required(envVariable: string, devFallback?: string): string {
+    const value = process.env[envVariable]
+    if (value !== undefined) {
+      return value
+    }
+    if (this.isDev && devFallback !== undefined) {
+      return devFallback
+    }
+    this.issues[envVariable] = 'missing'
+    return (undefined as unknown) as string
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  requiredJSON<T = any>(envVariable: string, devFallback?: T): T {
+    const value = process.env[envVariable]
+    if (value !== undefined) {
+      try {
+        return JSON.parse(value)
+      } catch {
+        this.issues[envVariable] = 'could not be parsed as JSON'
+        return (undefined as unknown) as T
+      }
+    }
+    if (this.isDev && devFallback !== undefined) {
+      return devFallback
+    }
+    this.issues[envVariable] = 'missing'
+    return (undefined as unknown) as T
+  }
+
+  optional(envVariable: string): string | undefined {
+    return process.env[envVariable]
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  optionalJSON<T = any>(envVariable: string): T | undefined {
+    const value = process.env[envVariable]
+    if (value === undefined) {
+      return undefined
+    }
+
+    try {
+      return JSON.parse(value)
+    } catch (err) {
+      this.issues[envVariable] = 'could not be parsed as JSON'
+      return (undefined as unknown) as T
```

Comment:
This should just return `undefined` without typecasts

Reply:
Not really. This is to provide better TypeScript types for the configuration, and allow the loader to run as far as it can so we can list all issues.